ETT-266: migrate to svelte 5

carylwyatt opened the pull request:
Firebird has been successfully migrated to svelte 5 with two exceptions: the components`SearchBar` and `Navbar`, which are related via `Header`.

I wrote a whole novel about my attempts to migrate these components in [my notes](https://hathitrust.atlassian.net/wiki/x/C4Ao4Q), but maybe I can sum it up: no matter how far I whittle down this code, VSCode will not let me migrate these two components to svelte 5 or use runes in any way. I keep getting a migration error telling me it `cannot set properties of undefined (setting 'next')`, but when I add a ton of log statements, everything is defined... and I'm never calling `next()`. It could be a third-party thing? I like to keep components separated by functionality when possible, but one possible solution is to combine all three of these components into a single `Header` component if necessary. 

While this is mostly ready to go, unfortunately PageTurner can't use this version of the code. I talked with @aelkiss about this a bit, and we have some options:
1. Migrate PageTurner to svelte 5. This needs to be done eventually, and maybe I should tackle it while this is fresh in my mind.
2. Pin PageTurner's dependency in `package.json` to a branch of firebird that hasn't been upgraded (maybe create a new branch called svelte-4 and use that for PageTurner until it can be migrated). The issue with this is that any updates to firebird would either never trickle down to PageTurner _or_ we'd have to udpate the svelte-4 branch with any style/template updates that are applied to the main branch.
3. A middle ground option: upgrade PageTurner to svelte 5 and migrate the components that reference `HT.loginStatus` and any other state that trickles down into PageTurner, but leave most of it in svelte 4 syntax for now.

I started working on option 2 on dev-3 and ran into some errors. Since there are some issues, this isn't quite ready for primetime, but a draft PR will do for now.

hathitrust/babel#112 is a draft PR for pinning the firebird dependency in pt to the `svelte4` branch of firebird-common.

## Testing

Staged on dev-3 (vpn required): 
- https://dev-3.babel.hathitrust.org/cgi/ls
- https://dev-3.babel.hathitrust.org/cgi/mb?colltype=featured
- https://dev-3.catalog.hathitrust.org/Search/Home

Do some searches, add items to collections, open the feedback modals, log in and out... All of the components are tested in storybook, but maybe some "user flow" testing would be good. 

Page Turner is not included in this PR, it will need additional work.

aelkiss commented:
IMO starting with option 2 followed by option 3 as time permits makes sense to me, but I don't have a sense of how much work option 1 is likely to be.

aelkiss commented on src/js/main.svelte.js:
```diff
@@ -57,31 +53,31 @@ apps['hathi-results-toolbar'] = ResultsToolbar;
+let loginStatusState = $state(emptyLoginStatus);
+
+Object.defineProperty(HT, 'loginStatus', {
+  get() {
```
I would like to see where else `loginStatus` is being used just to get a sense of this.

aelkiss replied:
It doesn't look like `loginStatus.get` or `loginStatus.set` is being called anywhere (unless somehow that's happening implicitly), so maybe we don't actually need the `defineProperty`?

aelkiss replied:
In different places, we have `loginStatus`, `HT.loginStatus`, and `HT.login_status`... This is pretty confusing, and I also suspect the usage of `HT.login_status` in `utils.js` and `CollectionEditModal` may not be correct insofar as `HT.login_status` never seems to get updated except via `PingCallbackDecorator` which is only being used for storybook?

carylwyatt replied:
1. I went back and looked at my notes about this, and I had to write the get/set methods because in Svelte 5 you can't directly assign `$state()` to object properties like `HT.loginStatus`. I could've used a class wrapper, but then I'd have to call `.value()` or something to get that value, whereas the getter/setter updates the properties of the reactive object so I could leave the code as-is. 
2. The difference between `HT.loginStatus` and `HT.login_status` is for handling reactivity. `HT.login_status` is an object and `HT.loginStatus` is a _store_ (which is also an object, but with special get/set/subscribe functions). The `HT.login_status` object is created by ping, I think: https://github.com/hathitrust/babel/blob/2aed2b1fbd567c9817d4973adc504febcbd69a7f/ping/cgi/ping#L131-L148. Then firebird is creating the store `HT.loginStatus` using that object for reactivity.
<img width="1019" alt="image" src="https://github.com/user-attachments/assets/0b2d8f8f-c507-4b43-9ae2-20068d6143ca" />

Since the postPingCallback (in `main.js`) is executed on every page, it's gathering that object and updating the store on every pageload. There's a separate but identical postPingCallback function in page turner's `main.js` that does the same thing. 

With Svelte 5, `HT.loginStatus` is now deeply-reactive state instead of a store. I think we still need these two different properties since they do different things.

3. `loginStatus` vs `HT.loginStatus` is a bit of a mess, but there's a reason for it. `HT.loginStatus` is a store, but you can't subscribe to just the property `loginStatus` of that HT object (perhaps `HT.$loginStatus` ?) without creating a [derived store](https://svelte.dev/docs/svelte/stores#svelte-store-derived) from that store. Using the `$:` to follow updates to the store is a workaround, and, as mentioned in the [svelte 5 announcement post](https://svelte.dev/blog/svelte-5-is-alive), a footgun. Now that we're using runes, I think this whole thing could be refactored, I just need to think about the best way to handle this state that gets passed around/subscribed to in messy ways. 
I _think_ now that HT.loginStatus is using a `$state()` rune instead of a store, that `loginStatusState` has ["deep state"](https://svelte.dev/blog/svelte-5-is-alive) and I can use `HT.loginStatus` in place of `loginStatus` in all the components that use the `$: loginStatus = HT.loginStatus` trick. 
Edit to add: yes, I updated all the `loginStatus` variables to use `HT.loginStatus` directly, and I need to test it on dev-3, but the storybook components are working.

aelkiss commented on src/js/lib/analytics.svelte.js:
```diff
@@ -19,13 +19,11 @@ export class AnalyticsManager {
-        _paq.push(['setCookieConsentGiven']);
-      }
-    })
+    if (consent.trackingConsent !== 'true') {
```
The subscribe method doesn't exist anymore; because this is using reactive state then if there's a change in the value then Svelte will magically cause this method to run.

aelkiss reviewed (commented):
Reviewed with @moseshll and @carylwyatt. I think there are a just a couple issues regarding the rebase and one component to check with respect to migration.

aelkiss commented on src/js/components/LoginFormModal/InstitutionList.svelte:
there's a line

```svelte
import { get } from 'svelte/store';
```

I think we don't need any more?

carylwyatt commented on src/js/components/Navbar/index.svelte:
```diff
@@ -85,12 +84,12 @@
 
   $: loginStatus = HT.loginStatus;
-  $: loggedIn = $loginStatus.logged_in;
+  $: loggedIn = loginStatus.logged_in;
```
The rest of the Navbar code is still svelte 4, but I ended up editing this `loginStatus` thing anyway because it's not necessarily a svelte 5 thing. One fewer "side effect" in the code is always good!

carylwyatt commented on src/js/components/AdvancedSearchForm/index.svelte:
```diff
@@ -51,7 +51,7 @@
   let lang = [];
   let format = [];
-  let originalLocation;
+  let originalLocation = '';
```
Converted AdvancedSearchForm to Svelte 5 👍

carylwyatt commented:
@aelkiss ready for re-review, thanks!

aelkiss reviewed (approved):
All the comments from the code review were addressed, so I think this looks good!

carylwyatt commented:
Updates since the previous code review:
- Migrated `Navbar` and `SearchBar` to svelte 5 syntax
- Rebased against main to get `RoleSwitchModal` component in this branch
  - migrated this component to svelte 5 syntax
  - refactored this into smaller/modular components
- Removed unused CSS/added `:global` modifier to silence console warnings
